api/src/main/java/org/pmiops/workbench/actionaudit/adapters/WorkspaceAuditAdapterImpl.kt

```diff
@@ -86,45 +116,42 @@ constructor(
     }
 
     override fun fireDuplicateAction(
-        sourceWorkspaceDbModel: DbWorkspace,
-        destinationWorkspaceDbModel: DbWorkspace
+        sourceWorkspaceId: Long,
+        destinationWorkspaceId: Long,
+        destinationWorkspace: Workspace
     ) {
         try {
             // We represent the duplication as a single action with events with different
             // ActionTypes: DUPLICATE_FROM and DUPLICATE_TO. The latter action generates many events
             // (similar to how CREATE is handled) for all the properties. I'm not (currently) filtering
             // out values that are the same.
-            val actionId = actionIdProvider.get()
-            val userId = userProvider.get().userId
-            val userEmail = userProvider.get().email
-            val timestamp = clock.millis()
             // TODO(jaycarlton): consider grabbing source event properties as well
+            val info = getAuditEventInfo()
             val sourceEvent = ActionAuditEvent(
-                    actionId = actionId,
-                    agentEmailMaybe = userEmail,
+                    actionId = info.actionId,
+                    agentEmailMaybe = info.userEmail,
                     actionType = ActionType.DUPLICATE_FROM,
                     agentType = AgentType.USER,
-                    agentId = userId,
+                    agentId = info.userId,
                     targetType = TargetType.WORKSPACE,
-                    targetIdMaybe = sourceWorkspaceDbModel.workspaceId,
-                    timestamp = timestamp
+                    targetIdMaybe = sourceWorkspaceId,
+                    timestamp = info.timestamp
             )
             val destinationPropertyValues = TargetPropertyExtractor.getPropertyValuesByName(
-                    WorkspaceTargetProperty.values(),
-                    WorkspaceConversionUtils.toApiWorkspace(destinationWorkspaceDbModel))
+                    WorkspaceTargetProperty.values(), destinationWorkspace)
 
             val destinationEvents: List<ActionAuditEvent> = destinationPropertyValues.entries
                     .map { ActionAuditEvent(
-                            actionId = actionId,
-                            agentEmailMaybe = userEmail,
+                            actionId = info.actionId,
```

Comment:
That class is purposely very vanilla, as each field is a JSON column, and they all share the same schema. I also removed from it any knowledge of specific AoU concepts or StackDriver ones.

There are a number of places where it might look like grouping or splitting things makes sense, but there are enough exceptions that I don't feel that makes sense.

Reply:
yeah, a more concrete example might look something like 

```
CreateWorkspaceActionAuditEvent(
  newWorkspaceId = dbWorkspaceId,
  fieldName = it.key,
  fieldValue = it.value,
  metadata = info)
```

a method call version of the above works too

```
factory.createWorkspaceActionAuditEvent(
  newWorkspaceId = dbWorkspaceId,
  fieldName = it.key,
  fieldValue = it.value,
  metadata = info)
```

Nice thing about either approach is that it makes it really clear what is unique about this event and you can also introduce a layer of abstraction with argument names that are more relevant.


I don't think it's necessary to do. Just a thought I had while reading through this.